### ISSUE TITLE
YOLOv3 odd detection behaviour vs YOLOv2

### ISSUE THREAD

**Issue by abagshaw** [taehoonlee/tensornets#17]
Here's the same [image](https://unsplash.com/photos/grKziVsagXs) put through `YOLOv2COCO` and then `YOLOv3COCO` respectively (threshold set at 0.4 and input resolution at 416x416 for both):

![YOLOv2](https://user-images.githubusercontent.com/8594541/39488635-bd882dfa-4d37-11e8-96b0-0aa1aabc4004.PNG)
![YOLOv3](https://user-images.githubusercontent.com/8594541/39488636-bdab695a-4d37-11e8-8453-97484154a9f6.PNG)

YOLOv3 does pick up the smaller people (as one would expect) - but for some reason it seems to predicting a much smaller bounding box than it should. I've done some experimenting and similar behavior is exhibited on quite a few images (sometimes the proper box appears and a smaller, inner one also appears - in this case, though, only the smaller one is appearing).

It's possible that this is just due to a poor prediction on the model's part, but my guess is that it isn't and due to some problem with the NMS function (almost as if it's doing non-minimal-suppression...if that's a thing 😄 ) or other post processing step.

**Comment by taehoonlee** [comment]
Yes, I suspect the post processing too, because I just had translated the architecture of the original YOLOv3 and concatenated the result boxes of three different scales (see [codes](https://github.com/taehoonlee/tensornets/blob/master/tensornets/references/yolo_utils.py#L74)). I think I should carefully investigate incorporation of the results from three-scales. Thank you for the comments, @abagshaw.

**Comment by taehoonlee** [comment]
@abagshaw, I got it. The keras implementation you mentioned [concatenates](https://github.com/experiencor/keras-yolo3/blob/master/yolo3_one_file_to_detect_them_all.py#L418) the result boxes of three different scales like TensorNets. But it does NMS after concat while TensorNets does concat after NMS. I'll compare the two approaches and let me know you.

**Comment by taehoonlee** [comment]
@abagshaw, I confirmed the softmax issue and will revise the post processing with respect to [the original one](https://github.com/pjreddie/darknet/blob/master/src/yolo_layer.c#L132). Thank you!

**Comment by taehoonlee** [comment]
@abagshaw, I revised the NMS issue first. The [updates](https://github.com/taehoonlee/tensornets/commit/93bd531be8fcd1c2fee56b0d3626a7a769005847) improved the mAP on VOC2007 by 2%.
Could you check to see if the smaller bbox problem is still happening?

**Comment by abagshaw** [comment]
@taehoonlee Problem solved! It's working beautifully now as far as I can see. Thanks so much!
